testing: cttest_close_frees_job fails on Linux

Producing the following output, rather than succeeding.  A build against master last succeeded at Mon Jul 15 01:22:02 UTC 2019:

    ct/_ctcheck
    ........................F................................................

    cttest_close_frees_job: failure
    start server port=40517 pid=4338
    >28 reserve-with-timeout 1
    >29 put 0 0 100 1
    >29 a
    <29 INSERTED 1
    <28 RESERVED 1 1
    <28 a
    >29 stats-job 1
    <29 OK 149
    <29 ---
    id: 1
    tube: default
    state: reserved
    pri: 0
    age: 0
    delay: 0
    ttr: 100
    time-left: 99
    file: 0
    reserves: 1
    timeouts: 0
    releases: 0
    buries: 0
    kicks: 0

    >29 stats-job 1
    <29 OK 149
    <29 ---
    id: 1
    tube: default
    state: reserved
    pri: 0
    age: 0
    delay: 0
    ttr: 100
    time-left: 99
    file: 0
    reserves: 1
    timeouts: 0
    releases: 0
    buries: 0
    kicks: 0

    testserv.c:260: test: strstr(line, sub)
    testserv.c:260: "
    state: ready
    " not in "---
    id: 1
    tube: default
    state: reserved
    pri: 0
    age: 0
    delay: 0
    ttr: 100
    time-left: 99
    file: 0
    reserves: 1
    timeouts: 0
    releases: 0
    buries: 0
    kicks: 0

    "

    1 failures; 0 errors.
    make[1]: *** [check] Error 1

What system are you running it on? What is the compiler version?

> On Jul 16, 2019, at 18:23, Alan Post <notifications@github.com> wrote:
> 
> Producing the following output, rather than succeeding. A build against master last succeeded at Mon Jul 15 01:22:02 UTC 2019:
> 
> ct/_ctcheck
> ........................F................................................
> 
> cttest_close_frees_job: failure
> start server port=40517 pid=4338
> >28 reserve-with-timeout 1
> >29 put 0 0 100 1
> >29 a
> <29 INSERTED 1
> <28 RESERVED 1 1
> <28 a
> >29 stats-job 1
> <29 OK 149
> <29 ---
> id: 1
> tube: default
> state: reserved
> pri: 0
> age: 0
> delay: 0
> ttr: 100
> time-left: 99
> file: 0
> reserves: 1
> timeouts: 0
> releases: 0
> buries: 0
> kicks: 0
> 
> >29 stats-job 1
> <29 OK 149
> <29 ---
> id: 1
> tube: default
> state: reserved
> pri: 0
> age: 0
> delay: 0
> ttr: 100
> time-left: 99
> file: 0
> reserves: 1
> timeouts: 0
> releases: 0
> buries: 0
> kicks: 0
> 
> testserv.c:260: test: strstr(line, sub)
> testserv.c:260: "
> state: ready
> " not in "---
> id: 1
> tube: default
> state: reserved
> pri: 0
> age: 0
> delay: 0
> ttr: 100
> time-left: 99
> file: 0
> reserves: 1
> timeouts: 0
> releases: 0
> buries: 0
> kicks: 0
> 
> "
> 
> 1 failures; 0 errors.
> make[1]: *** [check] Error 1
> make: *** [beanstalkd] Error 2
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

--- a/testserv.c
+++ b/testserv.c
@@ -19,8 +19,13 @@
 #include <errno.h>
 
 static int srvpid, port, fd, size;
-static int64 timeout = 5000000000LL; // 5s
 
+// Global timeout set for reading response in tests; 5sec.
+static int64 timeout = 5000000000LL;
+
+// Allocation pattern for wrapfalloc that replaces falloc in tests.
+// Zero value at N-th element means that N-th call to the falloc
+// should fail with ENOSPC result.
 static byte fallocpat[3];
 
 
@@ -130,6 +135,9 @@ kill_srvpid(void)
 
 #define SERVER() (progname=__func__, mustforksrv())
 
+// Forks the server storing the pid in srvpid.
+// The parent process returns port assigned.
+// The child process serves until the SIGTERM is received by it.
 static int
 mustforksrv(void)
 {
@@ -885,7 +893,7 @@ cttest_reserve_ttr_deadline_soon()
 }
 
 void
-cttest_close_frees_job()
+cttest_close_releases_job()
 {
     port = SERVER();
     int cons = mustdiallocal(port);
@@ -903,12 +911,14 @@ cttest_close_frees_job()
     ckrespsub(prod, "OK ");
     ckrespsub(prod, "\nstate: reserved\n");
 
-    // Closed consumer connection should make the job ready again.
+    // Closed consumer connection should make the job ready sooner than ttr=100.
     close(cons);
 
-    mustsend(prod, "stats-job 1\r\n");
-    ckrespsub(prod, "OK ");
-    ckrespsub(prod, "\nstate: ready\n");
+    // Job should be released in less than 1s. It is low expectation,
+    // but we do not make guarantees about how soon jobs should be released.
+    mustsend(prod, "reserve-with-timeout 1\r\n");
+    ckresp(prod, "RESERVED 1 1\r\n");
+    ckresp(prod, "a\r\n");
 }
 
 void

